src/transformers/models/ict/modeling_ict.py

```diff
@@ -0,0 +1,904 @@
+        self.blocks = nn.ModuleList([ICTBlock(config) for _ in range(config.num_hidden_layers)])
+
+        # Decoder head
+        self.layernorm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.token_embedding
+
+    @add_start_docstrings_to_model_forward(ICT_TRANSFORMER_INPUTS_DOCSTRING)
+    @add_code_sample_docstrings(
+        checkpoint=_CHECKPOINT_FOR_DOC,
+        output_type=BaseModelOutput,
+        config_class=_CONFIG_FOR_DOC,
+        modality="vision",
+        expected_output=_EXPECTED_OUTPUT_SHAPE,
+    )
+    def forward(
+        self,
+        pixel_values: Optional[torch.Tensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, BaseModelOutput]:
+        all_hidden_states = () if output_hidden_states else None
+        all_self_attentions = () if output_attentions else None
+
+        if pixel_values is None:
+            raise ValueError("You have to specify pixel_values")
+
+        pixel_values = pixel_values.to(torch.long)
+        _, t = pixel_values.size()
+
+        inputs_embeds = self.token_embedding(pixel_values)
+
+        if masks:
+            masks = masks.unsqueeze(2)
+            inputs_embeds = inputs_embeds * (1 - masks)
+
+        position_embeds = self.position_embedding[:, :t, :]
+        hidden_states = inputs_embeds + position_embeds
+        hidden_states = self.drop(hidden_states)
+
+        for _, block in enumerate(self.blocks):
+            if output_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
+
+            if self.gradient_checkpointing and self.training:
+
+                def create_custom_forward(module):
+                    def custom_forward(*inputs):
+                        return module(*inputs, output_attentions)
+
+                    return custom_forward
+
+                layer_outputs = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(block),
+                    hidden_states,
+                )
+            else:
+                layer_outputs = block(hidden_states, output_attentions)
+
+            hidden_states = layer_outputs[0]
+
+            if output_attentions:
+                all_self_attentions = all_self_attentions + (layer_outputs[1],)
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, all_hidden_states, all_self_attentions] if v is not None)
+
+        return BaseModelOutput(
+            last_hidden_state=hidden_states,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attentions,
+        )
+
+
+class BaseNetwork(nn.Module):
```

Comment:
From here to the end of `ICTGuidedUpsampler`, it's a cleaned up version of [models.py](https://github.com/raywzy/ICT/blob/main/Guided_Upsample/src/models.py), [networks.py](https://github.com/raywzy/ICT/blob/main/Guided_Upsample/src/networks.py), and [loss.py](https://github.com/raywzy/ICT/blob/main/Guided_Upsample/src/loss.py) in the original repo.

Reply:
Actually, I had something called [`ICTTransformerPreTrainedModel`](https://github.com/huggingface/transformers/blob/243f14260281cf1e117db39aaf41ab185e08bda0/src/transformers/models/ict/modeling_ict.py#L151-L179) but it looks like I can delete this `BaseNetwork` and rename `ICTTransformerPreTrainedModel` to `ICTPretrainedModel`. 

Then it looks like I can use that `ICTPretrainedModel` class everywhere instead of creating two `PretrainedModel` classes for `ICTTransformerPreTrainedModel` to be used like ```
class ICTTransformerModel(ICTTransformerPreTrainedModel)
``` 
and `ICTPretrainedGuidedUpsampler` to be used like:
```
class ICTPretrainedGuidedUpsampler(ICTPretrainedModel)
```
Let me know!

Reply:
Actually, never mind that. I feel like there should be a class that inherits `PreTrainedModel` _and_ also uses `ICTGuidedUpsamplerConfig` as its `config_class`. So something like this? 

```python
class ICTPretrainedGuidedUpsampler(PreTrainedModel):
    """
    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
    models.
    """

    config_class = ICTTransformerConfig
    base_model_prefix = "ict"
    main_input_name = "pixel_values"
    supports_gradient_checkpointing = True
    _no_split_modules = []

    def _init_weights(self, module: Union[nn.Linear, nn.Embedding, nn.LayerNorm]) -> None:
        """Initialize the weights"""
        if isinstance(module, (nn.Linear, nn.Embedding)):
            # Slightly different from the ViT version which uses truncated_normal for initialization
            # https://github.com/raywzy/ICT/blob/59dd12d374d47cdf0dce90923017ca3657e6aa0b/Transformer/models/model.py#L159-L166
            module.weight.data = nn.init.normal_(
                module.weight.data.to(torch.float32), mean=0.0, std=self.config.initializer_range
            ).to(module.weight.dtype)
            if isinstance(module, nn.Linear) and module.bias is not None:
                module.bias.data.zero_()
        elif isinstance(module, nn.LayerNorm):
            module.bias.data.zero_()
            module.weight.data.fill_(1.0)

    def _set_gradient_checkpointing(self, module, value: bool = False) -> None:
        if isinstance(module, ICTTransformerModel):
            module.gradient_checkpointing = value
```

Reply:
Actually again, I'll ask some high-level questions in this thread: https://huggingface.slack.com/archives/C04Q5M5U0UX/p1678968306113829